File: Residential.py
# Residential Controller 
import random
import time


 
# --- Classes ---

class CallButton():
    def __init__(self, direction, floor):
        self.direction = direction
        self.floor = floor
        self.light = 'OFF'

class Column():
    def __init__(self, ID, numberOfFloors, numberOfElevators):
        self.ID = ID
        self.elevatorList = []
        self.floorList = []
        self.callButtonList = []
                                                                                                                                                                                                                                                       
        for i in range (numberOfElevators):
            elevator = Elevator(i + 1, numberOfFloors)
            self.elevatorList.append(elevator)

        for i in range (numberOfFloors):
            self.floorList.append(i)

        for i in range (numberOfFloors):
            if i != 1 :
                callbutton = CallButton('DOWN', i)
                self.callButtonList.append(callbutton)
            if i != numberOfFloors :
                callbutton = CallButton('UP', i)
                self.callButtonList.append(callbutton)

class FloorRequestButton():
    def __init__(self, ID):
        self.ID = ID
        self.Pressed = False

class Elevator():
    def __init__(self, ID, numberOfFloors):
        self.ID = ID
        # self.State = 'IDLE'
        self.Position = 1
        self.Direction = 'UP'
        self.StopList = [] 
        self.FloorRequestButton = []
        self.Door = 'CLOSED'
        self.BufferDirection = 'UP'
        self.BufferList = []

        for i in range (numberOfFloors):
            self.FloorRequestButton.append(i)


class Controller():
    def __init__ (self, howManyColumns, howManyFloors, howManyElevatorsPerColumn):
        self.columnList = []
        for i in range(howManyColumns):
            self.columnList.append(Column(i +1, howManyFloors, howManyElevatorsPerColumn))


    def Listen(self):
        while True:
            try:
                x = int(input('Which floor are you calling from?   : '))
            except ValueError:
                print('ERROR::: enter a vlid number!')
            if (x >= 1 and x <= len(self.columnList[0].floorList)) :
                self.columnList[0].callButtonList[x].Position = int(x)
                break
            else:
                print('ERROR::: enter a vlid number!') 
            
        
        while True:
            try:
                dir = input('What is your direction(U/D)?  : ')
            except ValueError:
                print('ERROR::: enter a vlid number!')
            if (dir in ['u', 'U', 'up', 'UP', 'Up', 'd', 'D', 'down', 'Down', 'DOWN']):
                if (dir in ['u', 'U', 'up', 'UP', 'Up']):
                    self.columnList[0].callButtonList[x].Direction = 'UP'
                    break
                elif (dir in ['d', 'D', 'down', 'Down', 'DOWN']):
                    self.columnList[0].callButtonList[x].Direction = 'DOWN'
                    break
            else:
                print('ERROR::: enter a vlid direction (U/D)!')
            
        return self.columnList[0].callButtonList[x]
            
    def UpdateList(self, elList, Position, state):
        elList.append([Position, state])
        elList.sort()
    # finding the elevator which has the shortest stopList and is closer to the user position
    def FindClosestWithShortestListElevator(self, elevatorsList, userPosition):
        # print ('FindClosestWithShortestListElevator elevators: ', elevators)
        distance = len(self.columnList[0].floorList)
        best = elevatorsList[0]
        listLength = 10
        for elevator in elevatorsList:
            if abs(elevator.Position - userPosition) < distance:
                if len(elevator.StopList) <= listLength:
                    listLength = len(elevator.StopList)
                    best = elevator
                    distance = abs(elevator.Position - userPosition)
        return best
    def FindTheShortestStopList(self, elevatorsList):
        listLength = 10
        best = elevatorsList[0]
        for elevator in elevatorsList:
            if len(elevator.StopList) < listLength:
                listLength = len(elevator.StopList)
                best = elevator
        return best
    def RequestElevator(self, columnID, Position, Direction):
        #  print('user is here: ', Position, 'and going ',Direction)
        GoodElevators = []
        BadElevators = []
        print ('User is at ', Position, ' and is goind', Direction)
        for elevator in self.columnList[columnID - 1].elevatorList:
            print ('elvator', elevator.ID, ' is at floor ', elevator.Position, ' and its direction is ', elevator.Direction)
            time.sleep(.3)
            # the first condition (if) is the ideal case and rarely happens 
            if (elevator.Position == Position and elevator.Direction == Direction) :
                if (elevator.Door == 'OPEN'):
                    self.UpdateList(elevator.StopList, Position, 'IN')
                    print('the elevator', elevator.ID, ' is comming')
                    return elevator
            # Usually the following cases happen
            elif (Direction == elevator.Direction):
                if (elevator.Direction == 'UP' and elevator.Position < Position):
                    GoodElevators.append(elevator)
                    # print('el no', elevator.ID, 'added to GOOD1')
                elif (elevator.Direction == 'DOWN' and elevator.Position > Position):
                    GoodElevators.append(elevator)
                    # print('el no', elevator.ID, 'added to GOOD2')
                else:
                    BadElevators.append(elevator)
            # other probable case: Elevator is IDLE
            elif (elevator.Direction == 'IDLE'):
                GoodElevators.append(elevator)
                # print('el no', elevator.ID, 'added to GOOD3')
            # the worse case depening on elevator direction and position, elevator will be added to BADLIST
            else:
                BadElevators.append(elevator)
                # print('el no', elevator.ID, 'added to BAD2')

        if len(GoodElevators) >= 1:
            # print('GOOD elevators')
            # for i in range(len(GoodElevators)):
            #     print ('el no. ', GoodElevators[i].ID)
            bestElevator = self.FindClosestWithShortestListElevator(GoodElevators, Position)
            self.UpdateList(bestElevator.StopList, Position, 'IN')
            print('the elevator', bestElevator.ID, ' is comming')
            # print('the elevator', bestElevator.ID, ' is comming')
            return bestElevator
        else :
            # print('BAD elevators')
            # for i in range(len(BadElevators)):
            #     print ('el no. ', BadElevators[i].ID)
            bestElevator = self.FindTheShortestStopList(BadElevators)
            self.UpdateList(bestElevator.BufferList, Position, 'IN')
            bestElevator.BufferDirection = Direction
            print('the elevator', bestElevator.ID, ' is comming')
            # print('the badelevator', bestElevator.ID, ' is comming')
            return bestElevator
    
    def move(self, elevator):
        # print('elevator1: ', elevator)
        # print('length stoplist: ', len(elevator.StopList))
        while (len(elevator.StopList) > 0) :
            # print('stops: ', elevator.StopList) 
            time.sleep(3)
            if (elevator.StopList[0][0] > elevator.Position) :
                elevator.Direction = 'UP'
                while (elevator.Position < elevator.StopList[0][0]):
                    elevator.Position += 1
                    print('Elevator ', elevator.ID, ' is at Floor ', elevator.Position)
                    time.sleep(1)
                    if (elevator.Position == len(self.columnList)):
                        elevator.Direction = 'IDLE'
                elevator.Door = 'OPEN'
                print('Door is open')
                if (elevator.StopList[0][1] == 'IN'):
                    elevator.StopList.pop(0)
                    RequestedFloor = int(input('input your destionation floor: '))
                    controller.RequestFloor(elevator, RequestedFloor)
                else:
                    elevator.StopList.pop(0)
            else:
                elevator.Direction = 'DOWN'
                while (elevator.Position > elevator.StopList[0][0]):
                    elevator.Position -= 1
                    print('Elevator ', elevator.ID, ' is at Floor ', elevator.Position)
                    time.sleep(1)
                    if (elevator.Position == 1):
                        elevator.Direction = 'IDLE'
                elevator.Door = 'OPEN'
                print('Door is open')
                if (elevator.StopList[0][1] == 'IN'):
                    elevator.StopList.pop(len(elevator.StopList) - 1)
                    RequestedFloor = int(input('input your destionation floor: '))
                    # print('requestedFloor: ', RequestedFloor)
                    controller.RequestFloor(elevator, RequestedFloor)
                else:
                    elevator.StopList.pop(len(elevator.StopList) - 1)
            time.sleep(1)
            elevator.Door = 'CLOSED'
            print('Door is closed')
            elevator.Direction = 'IDLE'

        if (len(elevator.BufferList) > 0) :
            # print('stoplist: ', elevator.StopList)
            # print('bufferlist: ', elevator.BufferList)
            elevator.StopList = elevator.BufferList
            # for i in range(len(elevator.BufferList) - 1) :
            #     elevator.StopList[i] = elevator.BufferList[i]
            #     print('stop ', elevator.StopList[i], 'buffer ', elevator.BufferList[i])
            elevator.Direction = elevator.BufferDirection
            elevator.BufferList = []
            # print('stoplist: ', elevator.StopList)
            # print('stoplist: ', elevator.BufferList)
            self.move(elevator)
        else :
            elevator.direction = 'IDLE'
        # else :
            # while (elevator.Position <= len(column.floorList) or len(elevator.BufferList) > 0):
            #         if elevator.BufferList == 'UP' :
            #             elevator.Position += 1
            #         else :
            #             elevator.Position -= 1
            #         print('Elevator ', elevator.ID, ' is at ', elevator.Position, 'Floor')
            #         time.sleep(1)
            # elevator.Door = 'OPEN'
            # print('Door is open')
            # time.sleep(1)
            # elevator.Door = 'CLOSED'
            # print('Door is closed')

    def RequestFloor(self, elevator, RequestedFloor):
        # print('requestedFloor: ', RequestedFloor, 'elevatorID: ', elevator.ID, elevator.StopList)
        if (elevator.Direction != 'DOWN' and RequestedFloor > elevator.Position) or (elevator.Direction != 'UP' and RequestedFloor < elevator.Position) or (len(elevator.StopList) == 0):
            self.UpdateList(elevator.StopList, RequestedFloor, 'OUT')
            # print('elevatorID: ', elevator.ID, elevator.StopList)
            return
        else :
            return 
# --- /Classes ---



# # --- NO SCENARIO - AUTOMATED VERSION ---
# controller = Controller(1, 10, 3)
# # --- Initialization of the elevators --- 
# for j in range (len(controller.columnList)):
#     for i in range (len(controller.columnList[j].elevatorList)) :
#         controller.columnList[j].elevatorList[i].Position = random.randint(1, len(controller.columnList[j].floorList))
#         controller.columnList[j].elevatorList[i].StopList.append([random.randint(1, 10), 'OUT'])
#         if (controller.columnList[j].elevatorList[i].StopList[0][0] > controller.columnList[j].elevatorList[i].Position) :
#             controller.columnList[j].elevatorList[i].Direction = 'UP'
#         else : 
#             controller.columnList[j].elevatorList[i].Direction = 'DOWN'
#         random.seed((i + 10) * 10)
# # --- /Initialization of the elevators --- 
# callButton = controller.Listen()
# commingElevator = controller.RequestElevator(1, callButton.Position, callButton.Direction)
# controller.move(commingElevator)
# # --- /NO SCENARIO - AUTOMATED VERSION ---

# # # --- Scenario 1 ---
# controller = Controller(1, 10, 2)

# controller.columnList[0].elevatorList[0].Position = 2
# controller.columnList[0].elevatorList[0].Direction = 'IDLE'
# controller.columnList[0].elevatorList[1].Position = 6
# controller.columnList[0].elevatorList[1].Direction = 'IDLE'
# userPosition = 3
# userDirection = 'UP'
# userDestination = 7

# controller.columnList[0].callButtonList[userPosition].Position = userPosition
# controller.columnList[0].callButtonList[userPosition].Direction = userDirection

# print('******************* Scenario 1  *******************')
# print('******************* USER-1 goes from floor 3 to floor 7  *******************')
# commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
# controller.move(commingElevator)

# # # --- /Scenario 1 ---

# # --- Scenario 2 ---
controller = Controller(1, 10, 2)

controller.columnList[0].elevatorList[0].Position = 10
controller.columnList[0].elevatorList[0].Direction = 'IDLE'
controller.columnList[0].elevatorList[1].Position = 3
controller.columnList[0].elevatorList[1].Direction = 'IDLE'

print('****** 1st user at Floor 1 ******')
userPosition = 1
userDirection = 'UP'
userDestination = 6

controller.columnList[0].callButtonList[userPosition].Position = userPosition
controller.columnList[0].callButtonList[userPosition].Direction = userDirection
print('******************* Scenario 2 *******************');
print('******************* USER-1 goes from floor 3 to floor 6  *******************')
commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
controller.move(commingElevator)
print('****** 2nd user at Floor 3 ******')
print('****** 2nd user at Floor 3 ******')
userPosition = 3
userDirection = 'UP'
userDestination = 5

controller.columnList[0].callButtonList[userPosition].Position = userPosition
controller.columnList[0].callButtonList[userPosition].Direction = userDirection
print('******************* USER-2 goes from floor 3 to floor 5  *******************')
commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
controller.move(commingElevator)

print('****** 3rd user at Floor 9 ******')
print('****** 3rd user at Floor 9 ******')
userPosition = 9
userDirection = 'DOWN'
userDestination = 2
print('******************* USER-3 goes from floor 9 to floor 2  *******************')
controller.columnList[0].callButtonList[userPosition].Position = userPosition
controller.columnList[0].callButtonList[userPosition].Direction = userDirection

commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
controller.move(commingElevator)

# # --- /Scenario 2 ---


# # --- Scenario 3 ---

# controller = Controller(1, 10, 2)

# controller.columnList[0].elevatorList[0].Position = 10
# controller.columnList[0].elevatorList[0].Direction = 'IDLE'
# controller.columnList[0].elevatorList[1].Position = 3
# controller.columnList[0].elevatorList[1].Direction = 'UP'
# controller.columnList[0].elevatorList[1].StopList.append([6,'OUT'])

# print('****** 1st user at Floor 1 ******')
# userPosition = 3
# userDirection = 'DOWN'
# userDestination = 2

# controller.columnList[0].callButtonList[userPosition].Position = userPosition
# controller.columnList[0].callButtonList[userPosition].Direction = userDirection
# print('******************* Scenario 3 *******************')
# print('******************* USER-1 goes from floor 3 to floor 2  *******************')
# commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
# controller.move(commingElevator)
# # moving eleavtor 2 from floor 3 to floor 6
# controller.move(controller.columnList[0].elevatorList[1])

# print('****** 2nd user at Floor 10 ******')
# print('****** 2nd user at Floor 10 ******')
# userPosition = 10
# userDirection = 'DOWN'
# userDestination = 3
# print('******************* USER-2 goes from floor 10 to floor 3  *******************')
# controller.columnList[0].callButtonList[userPosition].Position = userPosition
# controller.columnList[0].callButtonList[userPosition].Direction = userDirection

# commingElevator = controller.RequestElevator(1, controller.columnList[0].callButtonList[userPosition].Position, controller.columnList[0].callButtonList[userPosition].Direction)
# controller.move(commingElevator)

# # --- /Scenario 3 ---

File: Residential.js

// --- Classes ----
class CallButton {
    constructor(direction, floor){
        this.direction = direction;
        this.floor = floor;
        this.light = 'OFF'
    }
}
class Column {
    constructor(ID, numberOfFloors, numberOfElevators){
        this.ID = ID;
        this.elevatorList = [];
        this.floorList = [];
        this.callButtonList = [];
        
        for (var i = 1; i <= numberOfElevators; i++) {
            // var elevator = 
            this.elevatorList.push(new Elevator(i , numberOfFloors))
        }
        for (var i = 1; i <= numberOfFloors; i++) {
            this.floorList.push(i)
        }
        for (var i = 1; i <= numberOfFloors; i++) {
            if (i != 1 ){
                var callbutton = new CallButton('DOWN', i)
                this.callButtonList.push(callbutton)
            }
            if (i != numberOfFloors ){
                callbutton = new CallButton('UP', i)
                this.callButtonList.push(callbutton)
            }
        }    
    }
}
class FloorRequestButton {
    constructor(ID){
        this.ID = ID;
        this.Pressed = false;
    }
}
class Elevator {
    constructor(ID, numberOfFloors){
        this.ID = ID;
        this.Position = 1;
        this.Direction = 'UP';
        this.StopList = [];
        this.FloorRequestButton = [];
        this.Door = 'CLOSED';
        this.BufferDirection = 'UP';
        this.BufferList = [];

        for (var i = 1; i <= numberOfFloors; i++){
            this.FloorRequestButton.push(i);
        }
    }
}
class Controller {
    constructor(howManyColumns, howManyFloors, howManyElevatorsPerColumn){
        this.columnList = [];

        for (var i = 1; i <= howManyColumns; i++){
            var column = new Column(i, howManyFloors, howManyElevatorsPerColumn);
            this.columnList.push(column);
        }
    }

    UpdateList(List, Position){
        List.push(Position);
        List.sort(function(a, b){return a-b});
    }
    // finding the elevator which has the shortest stopList and is closer to the user position
    FindClosestWithShortestListElevator (elevatorsList, userPosition){
        var distance = this.columnList[0].floorList.length;
        var best;
        var listLength = this.columnList[0].floorList.length;
        for (var elevator of elevatorsList){
            if (Math.abs(elevator.Position - userPosition) < distance){
                if (elevator.StopList.length <= listLength){
                    listLength = elevator.StopList.length;
                    distance = Math.abs(elevator.Position - userPosition);
                    best = elevator;
                }
            }
        }
        return best;
    }
    FindTheShortestStopList(elevatorsList){
        var listLength = 10;
        var best = elevatorsList[0];
        for (elevator of elevatorsList){
            var l = elevator.StopList.length
            if (l <= listLength){
                listLength = elevator.StopList.length;
                best = elevator;
            }
        }
        return best;
    }
    RequestElevator(RequestedFloor, Direction){
        var GoodElevators = [];
        var BadElevators = [];
        console.log('User is at ' + RequestedFloor + ' and is going ' + Direction);
        for (var elevator of this.columnList[0].elevatorList){
            console.log('elvator ' + elevator.ID + ' is at ' + elevator.Position + ' floor and its direction is ' + elevator.Direction);
            if (elevator.Position == RequestedFloor && elevator.Direction == Direction){
                if (elevator.Door == 'OPEN'){
                    this.UpdateList(elevator.StopList, RequestedFloor);
                    console.log('the best elevator No:' + elevator.ID + ' is comming');
                    return elevator;
                }
            }
            else if (Direction == elevator.Direction){
                if (elevator.Direction == 'UP' && elevator.Position < RequestedFloor){
                    GoodElevators.push(elevator);
                }
                else if ( elevator.Direction == 'DOWN' && elevator.Position > RequestedFloor){
                    GoodElevators.push(elevator);
                }
                else {
                    BadElevators.push(elevator);
                }
            }
            else if (elevator.Direction == 'IDLE'){
                GoodElevators.push(elevator);
            }
            else {
                BadElevators.push(elevator);
            }
        }
        if (GoodElevators.length >= 1) {
            var bestElevator = this.FindClosestWithShortestListElevator(GoodElevators, RequestedFloor);
            this.UpdateList(bestElevator.StopList, RequestedFloor);
            console.log('the elevator ' + bestElevator.ID + ' is comming');
            this.move(bestElevator);
            return bestElevator;
        }
        else {
            var bestElevator = this.FindTheShortestStopList(BadElevators);
            this.UpdateList(bestElevator.BufferList, RequestedFloor);
            bestElevator.BufferDirection = Direction;
            console.log('the elevator ' + bestElevator.ID + 'is comming');
            this.move(bestElevator);
            return bestElevator;
        }
    }
    move(elevator){
        while (elevator.StopList.length > 0){
            if (elevator.StopList[0] > elevator.Position){
                elevator.Direction = 'UP';
                while (elevator.Position < elevator.StopList[0]){
                    elevator.Position += 1;
                    console.log('Elevator ' + elevator.ID + ' is at floor ' + elevator.Position + 'Floor');
                    if (elevator.Position == this.columnList.length){
                        elevator.Direction = 'IDLE';
                    }
                }
                elevator.Door = 'OPEN';
                console.log('Door is open');
                elevator.StopList.splice(0,1);
            }
            else {
                elevator.Direction = 'DOWN';
                while (elevator.Position > elevator.StopList[0]){
                    elevator.Position -= 1;
                    console.log('Elevator ' + elevator.ID + ' is at floor ' + elevator.Position);
                    if (elevator.Position == 1){
                        elevator.Direction = 'IDLE';
                    }
                }
                elevator.Door = 'OPEN';
                console.log('Door is open');
                elevator.StopList.pop();
            }
            
            elevator.Door = 'CLOSED';
            console.log('Door is closed');
            elevator.Direction = 'IDLE';
        }
        if (elevator.BufferList.length > 0){
            elevator.StopList = eleavtor.BufferList;
            elevator.Direction = elevator.BufferDirection;
            elevator.BufferList = [];
            this.move(elevator);
        }
        else {
            elevator.Direction = 'IDLE';
        }
    }
    RequestFloor(elevator, RequestedFloor){
        
        this.UpdateList(elevator.StopList, RequestedFloor);
        this.move(elevator);
        return;
    }   
}
// --- /Classes ----   


// --- Scenarios ---
function Scenario1(){
    console.log('******************* ******************* *******************');
    console.log('*******************      Scenario 1     *******************');
    console.log('******************* ******************* *******************');
    controller1 = new Controller(1, 10, 2);

    controller1.columnList[0].elevatorList[0].Position = 2;
    controller1.columnList[0].elevatorList[0].Direction = 'IDLE';
    controller1.columnList[0].elevatorList[1].Position = 6;
    controller1.columnList[0].elevatorList[1].Direction = 'IDLE';
    console.log('******************* USER-1 goes from floor 3 to floor 7  *******************');
    RequestedFloor = 3;
    Direction = 'UP';
    Destination = 7;

    elevator = controller1.RequestElevator(RequestedFloor, Direction);
    controller1.RequestFloor(elevator, Destination);
}
function Scenario2(){
    console.log('******************* ******************* *******************');
    console.log('*******************      Scenario 2     *******************');
    console.log('******************* ******************* *******************');
    controller2 = new Controller(1, 10, 2);

    controller2.columnList[0].elevatorList[0].Position = 10;
    controller2.columnList[0].elevatorList[0].Direction = 'IDLE';
    controller2.columnList[0].elevatorList[1].Position = 3;
    controller2.columnList[0].elevatorList[1].Direction = 'IDLE';


    console.log('******************* USER-1 goes from floor 1 to floor 6  *******************');
    RequestedFloor = 1;
    Direction = 'UP';
    Destination = 6;


    elevator = controller2.RequestElevator(RequestedFloor, Direction);
    controller2.RequestFloor(elevator, Destination);

    console.log('******************* USER-2 goes from floor 3 to floor 5  *******************');
    RequestedFloor = 3;
    Direction = 'UP';
    Destination = 5;


    elevator = controller2.RequestElevator(RequestedFloor, Direction);
    controller2.RequestFloor(elevator, Destination);

    console.log('******************* USER-3 goes from floor 9 to floor 2  *******************');
    RequestedFloor = 9;
    Direction = 'DOWN';
    Destination = 2;


    elevator = controller2.RequestElevator(RequestedFloor, Direction);
    controller2.RequestFloor(elevator, Destination);
}
function Scenario3(){
    console.log('******************* ******************* *******************');
    console.log('*******************      Scenario 3     *******************');
    console.log('******************* ******************* *******************');
    controller3 = new Controller(1, 10, 2);

    controller3.columnList[0].elevatorList[0].Position = 10;
    controller3.columnList[0].elevatorList[0].Direction = 'IDLE';
    controller3.columnList[0].elevatorList[1].Position = 3;
    controller3.columnList[0].elevatorList[1].Direction = 'UP';
    controller3.columnList[0].elevatorList[1].StopList.push(6);


    console.log('******************* USER-1 goes from floor 3 to floor 2  *******************');
    RequestedFloor = 3;
    Direction = 'DOWN';
    Destination = 2;

    elevator = controller3.RequestElevator(RequestedFloor, Direction);
    controller3.RequestFloor(elevator, Destination);

    console.log('moving eleavtor 2 from floor 3 to floor 6');
    controller3.move(controller3.columnList[0].elevatorList[1]);

    console.log('******************* USER-2 goes from floor 10 to floor 3  *******************');
    RequestedFloor = 10;
    Direction = 'DOWN';
    Destination = 3;

    elevator = controller3.RequestElevator(RequestedFloor, Direction);
    controller3.RequestFloor(elevator, Destination);
}
//  --- /Scenarios---

// --- Main PRogram ---
Scenario1();
// Scenario2();
// Scenario3();
// --- /Main PRogram ---
